File: testdata/jennies/builders/references/TypescriptBuilder/src/some_pkg/person_builder_gen.ts
import * as cog from '../cog';
import * as some_pkg from '../some_pkg';
import * as other_pkg from '../other_pkg';

export class PersonBuilder implements cog.Builder<some_pkg.Person> {
    private readonly internal: some_pkg.Person;

    constructor() {
        this.internal = some_pkg.defaultPerson();
    }

    build(): some_pkg.Person {
        return this.internal;
    }

    name(name: other_pkg.Name): this {
        this.internal.name = name;
        return this;
    }
}

File: testdata/jennies/builders/references/GoBuilder/somepkg/person_builder_gen.go
package somepkg

import (
	cog "github.com/grafana/cog/generated/cog"
	otherpkg "github.com/grafana/cog/generated/otherpkg"
)

var _ cog.Builder[Person] = (*PersonBuilder)(nil)

type PersonBuilder struct {
    internal *Person
    errors map[string]cog.BuildErrors
}

func NewPersonBuilder() *PersonBuilder {
	resource := &Person{}
	builder := &PersonBuilder{
		internal: resource,
		errors: make(map[string]cog.BuildErrors),
	}

	builder.applyDefaults()

	return builder
}

func (builder *PersonBuilder) Build() (Person, error) {
	var errs cog.BuildErrors

	for _, err := range builder.errors {
		errs = append(errs, cog.MakeBuildErrors("Person", err)...)
	}

	if len(errs) != 0 {
		return Person{}, errs
	}

	return *builder.internal, nil
}

func (builder *PersonBuilder) Name(name otherpkg.Name) *PersonBuilder {
    builder.internal.Name = name

    return builder
}

func (builder *PersonBuilder) applyDefaults() {
}
